packages/tangle/scheduler.go

```diff
@@ -240,31 +297,74 @@ func (s *Scheduler) Clear() {
 	for q := s.buffer.Current(); q != nil; q = s.buffer.Next() {
 		s.buffer.RemoveNode(q.NodeID())
 		for _, id := range q.IDs() {
-			s.Events.MessageDiscarded.Trigger(MessageID(id))
+			messageID := MessageID(id)
+			s.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
+				messageMetadata.SetDiscardedTime(time.Now())
+			})
+			s.Events.MessageDiscarded.Trigger(messageID)
 		}
 	}
 }
 
+// isEligible returns true if the given messageID has either been scheduled or confirmed.
+func (s *Scheduler) isEligible(messageID MessageID) (eligible bool) {
+	s.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
+		eligible = messageMetadata.Scheduled() ||
+			s.tangle.ConfirmationOracle.IsMessageConfirmed(messageID)
+	})
+	return
+}
+
+// isReady returns true if the given messageID's parents are eligible.
+func (s *Scheduler) isReady(messageID MessageID) (ready bool) {
+	ready = true
+	s.tangle.Storage.Message(messageID).Consume(func(message *Message) {
+		message.ForEachParent(func(parent Parent) {
+			if !s.isEligible(parent.ID) { // parents are not eligible
+				ready = false
+				return
+			}
+		})
+	})
+
+	return
+}
+
+// tryReady tries to set the given message as ready.
+func (s *Scheduler) tryReady(messageID MessageID) {
+	if s.isReady(messageID) {
+		if err := s.Ready(messageID); err != nil {
+			s.Events.Error.Trigger(errors.Errorf("failed to mark %s as ready: %w", messageID, err))
+		}
+	}
+}
+
+// updateApprovers iterates over the direct approvers of the given messageID and
+// tries to mark them as ready.
+func (s *Scheduler) updateApprovers(messageID MessageID) {
+	s.tangle.Storage.Approvers(messageID).Consume(func(approver *Approver) {
+		s.tryReady(approver.approverMessageID)
+	})
+}
+
 func (s *Scheduler) submit(message *Message) error {
 	if s.stopped.IsSet() {
 		return ErrNotRunning
 	}
 
-	nodeID := identity.NewID(message.IssuerPublicKey())
-	mana := s.tangle.Options.SchedulerParams.AccessManaRetrieveFunc(nodeID)
-	if mana < MinMana {
-		s.Events.MessageDiscarded.Trigger(message.ID())
-		return schedulerutils.ErrInsufficientMana
-	}
-
-	err := s.buffer.Submit(message, mana)
-	if err != nil {
-		s.Events.MessageDiscarded.Trigger(message.ID())
-	}
-	if errors.Is(err, schedulerutils.ErrInboxExceeded) {
-		s.Events.NodeBlacklisted.Trigger(nodeID)
+	s.tangle.Storage.MessageMetadata(message.ID()).Consume(func(messageMetadata *MessageMetadata) {
+		// shortly before submitting we set the queued time
+		messageMetadata.SetQueuedTime(time.Now())
```

Comment:
Should here - and for `SetDiscardedTime` - `clock.SyncedTime()` be used? Or are there some (performance) reasons against this?

Reply:
No, that's an error. Good catch!